builtInFunctions/esdtDataStorage.go

```diff
@@ -417,6 +426,38 @@ func (e *esdtDataStorage) saveESDTMetaDataToSystemAccount(
 	return e.marshalAndSaveData(systemAcc, esdtDataOnSystemAcc, esdtNFTTokenKey)
 }
 
+func (e *esdtDataStorage) saveMetadataIfRequired(
+	esdtNFTTokenKey []byte,
+	systemAcc vmcommon.UserAccountHandler,
+	currentSaveData []byte,
+	esdtData *esdt.ESDigitalToken,
+) error {
+	if !e.enableEpochsHandler.IsAlwaysSaveTokenMetaDataEnabled() {
+		return nil
+	}
+	if len(currentSaveData) == 0 {
+		// optimization: do not try to write here the token metadata, it will be written automatically by the next step
+		return nil
+	}
+
+	esdtDataOnSystemAcc := &esdt.ESDigitalToken{}
+	err := e.marshaller.Unmarshal(esdtDataOnSystemAcc, currentSaveData)
+	if err != nil {
+		return err
+	}
+	if len(esdtDataOnSystemAcc.Reserved) > 0 {
+		return nil
+	}
+
+	esdtDataOnSystemAcc.TokenMetaData = esdtData.TokenMetaData
+	marshaledData, err := e.marshaller.Marshal(esdtDataOnSystemAcc)
```

Comment:
looks ok.  thinking about edge case - but this basically means, that if tokenMetaData is sent for old old tokens we just update the system account information. Looks ok.

Reply:
Great thanks.